hal/device.c

```diff
@@ -0,0 +1,52 @@
+#include "device.h"
+
+static struct device_type *devices;
+
+static struct device_type **find_device(const char *name, unsigned len) {
+  struct device_type **p;
+  for (p = &devices; *p; p = &(*p)->next)
+    if (strlen((*p)->name) == len && strncmp((*p)->name, name, len) == 0) break;
+  return p;
+}
+
+device_t *get_device(const char *type) {
```

Comment:
It is confusing to have `get_device` and `get_device_id`. You have to distinguish.